cocos/deprecated/CCArray.cpp

```diff
@@ -350,7 +350,8 @@ __Array* __Array::clone() const
     Ref* obj = nullptr;
     Ref* tmpObj = nullptr;
     Clonable* clonable = nullptr;
-    CCARRAY_FOREACH(this, obj)
+    const __Array* self( this );
+    CCARRAY_FOREACH(self, obj)
```

Comment:
The expansion of `CCARRAY_FOREACH` will begin with `if (this)` but since GCC 6 `this` is expected not to be null. Here is a quote from [GCC's release notes](https://gcc.gnu.org/gcc-6/changes.html):

> Value range propagation now assumes that the this pointer of C++ member functions is non-null.

Thus this test triggers the `-Wnonnull-compare` warning of GCC. Using another pointer (`self`) in the test prevents the warning.